tests/lean/run/2905.lean

```diff
@@ -0,0 +1,45 @@
+import Lean
+/-!
+# Only direct parents of classes are instances
+
+https://github.com/leanprover/lean4/issues/2905
+-/
+
+set_option structure.strictResolutionOrder true
+
+class A
+class B
+class C extends A
+class D extends A, B
+class E extends C, D
+
+/-!
+These were or were not instances before #5902 and still are or are not.
+-/
+/-- info: some 1000 -/
+#guard_msgs in #eval return (←Lean.Meta.getInstancePriority? `E.toC)
+/-- info: some 1000 -/
+#guard_msgs in #eval return (←Lean.Meta.getInstancePriority? `E.toD)
+/-- info: none -/
+#guard_msgs in #eval return (←Lean.Meta.getInstancePriority? `E.toA)
+
+/-!
+This was an instance before #5902 and no longer is.
+-/
+/-- info: none -/
+#guard_msgs in #eval return (←Lean.Meta.getInstancePriority? `E.toB)
+
+
+/-!
+Check that `A` is not an instance, since it is implied by the others.
+-/
+class E' extends C, D, A
```

Comment:
This one surprises me a little; shouldn't it be an instance because the user asked for it in `extends`?

Reply:
(unless this emits a warning, in which case I have no objection besides the test not checking for that warning)

Reply:
No, there's no warning. This feature is necessary for being able to satisfy the strict C3 linearization algorithm in some cases without affecting what the actual instances are. Currently strict mode is not the default, but I hope we can refactor mathlib at some point to have a strict linearization.